// @ironclad/rivet-core/dist/esm/model/nodes/AudioNode.js
import { NodeImpl, nodeDefinition } from '../NodeImpl.js';
import { nanoid } from 'nanoid/non-secure';
import { base64ToUint8Array, expectType, } from '../../index.js';
export class AudioNodeImpl extends NodeImpl {
    static create() {
        return {
            id: nanoid(),
            type: 'audio',
            title: 'Audio',
            visualData: { x: 0, y: 0, width: 300 },
            data: {
                useDataInput: false,
            },
        };
    }
    getInputDefinitions() {
        const inputDefinitions = [];
        if (this.chartNode.data.useDataInput) {
            inputDefinitions.push({
                id: 'data',
                title: 'Data',
                dataType: 'string',
            });
        }
        return inputDefinitions;
    }
    getOutputDefinitions() {
        return [
            {
                id: 'data',
                title: 'Audio Data',
                dataType: 'audio',
            },
        ];
    }
    getEditors() {
        return [
            {
                type: 'fileBrowser',
                label: 'Audio File',
                dataKey: 'data',
                useInputToggleDataKey: 'useDataInput',
                accept: 'audio/*',
            },
        ];
    }
    static getUIData() {
        return {
            contextMenuTitle: 'Audio',
            group: 'Data',
            infoBoxTitle: 'Audio Node',
            infoBoxBody: 'Defines an audio sample for use with other nodes. Can convert a binary type into an audio type.',
        };
    }
    async process(inputData, context) {
        let data;
        if (this.chartNode.data.useDataInput) {
            data = expectType(inputData['data'], 'binary');
        }
        else {
            const dataRef = this.data.data?.refId;
            if (!dataRef) {
                throw new Error('No data ref');
            }
            const encodedData = context.project.data?.[dataRef];
            if (!encodedData) {
                throw new Error(`No data at ref ${dataRef}`);
            }
            data = base64ToUint8Array(encodedData);
        }
        return {
            ['data']: {
                type: 'audio',
                value: { data },
            },
        };
    }
}
export const audioNode = nodeDefinition(AudioNodeImpl, 'Audio');
