// @ironclad/rivet-core/dist/esm/model/nodes/ExtractObjectPathNode.js
import { nanoid } from 'nanoid/non-secure';
import { NodeImpl, nodeDefinition } from '../NodeImpl.js';
import { JSONPath } from 'jsonpath-plus';
import { expectType } from '../../utils/expectType.js';
import { coerceTypeOptional } from '../../index.js';
import { dedent } from 'ts-dedent';
export class ExtractObjectPathNodeImpl extends NodeImpl {
    static create() {
        const chartNode = {
            type: 'extractObjectPath',
            title: 'Extract Object Path',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 250,
            },
            data: {
                path: '$',
                usePathInput: false,
            },
        };
        return chartNode;
    }
    getInputDefinitions() {
        const inputDefinitions = [
            {
                id: 'object',
                title: 'Object',
                dataType: 'object',
                required: true,
            },
        ];
        if (this.chartNode.data.usePathInput) {
            inputDefinitions.push({
                id: 'path',
                title: 'Path',
                dataType: 'string',
                required: true,
            });
        }
        return inputDefinitions;
    }
    getOutputDefinitions() {
        return [
            {
                id: 'match',
                title: 'Match',
                dataType: 'any',
            },
            {
                id: 'all_matches',
                title: 'All Matches',
                dataType: 'any[]',
            },
        ];
    }
    getEditors() {
        return [
            {
                type: 'code',
                label: 'Path',
                dataKey: 'path',
                language: 'jsonpath',
                useInputToggleDataKey: 'usePathInput',
            },
        ];
    }
    getBody() {
        return this.data.usePathInput ? '(Using Input)' : this.data.path;
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Extracts the value at the specified path from the input value. The path uses JSONPath notation to navigate through the value.
      `,
            infoBoxTitle: 'Extract Object Path Node',
            contextMenuTitle: 'Extract Object Path',
            group: ['Objects'],
        };
    }
    async process(inputs) {
        const inputObject = coerceTypeOptional(inputs['object'], 'object');
        const inputPath = this.chartNode.data.usePathInput
            ? expectType(inputs['path'], 'string')
            : this.chartNode.data.path;
        if (!inputPath) {
            throw new Error('Path input is not provided');
        }
        let matches;
        try {
            matches = JSONPath({ json: inputObject ?? null, path: inputPath.trim() });
        }
        catch (err) {
            matches = [];
        }
        if (matches.length === 0) {
            return {
                ['match']: {
                    type: 'control-flow-excluded',
                    value: undefined,
                },
                ['all_matches']: {
                    type: 'any[]',
                    value: [],
                },
            };
        }
        return {
            ['match']: {
                type: 'any',
                value: matches[0],
            },
            ['all_matches']: {
                type: 'any[]',
                value: matches,
            },
        };
    }
}
export const extractObjectPathNode = nodeDefinition(ExtractObjectPathNodeImpl, 'Extract Object Path');
