// @ironclad/rivet-core/dist/esm/model/nodes/GptFunctionNode.js
import { nanoid } from 'nanoid/non-secure';
import { NodeImpl, nodeDefinition } from '../NodeImpl.js';
import { dedent } from 'ts-dedent';
export class GptFunctionNodeImpl extends NodeImpl {
    static create() {
        const chartNode = {
            type: 'gptFunction',
            title: 'GPT Function',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 250,
            },
            data: {
                name: 'newFunction',
                description: 'No description provided',
                schema: `{
  "type": "object",
  "properties": {}
}`,
            },
        };
        return chartNode;
    }
    getInputDefinitions() {
        return [];
    }
    getOutputDefinitions() {
        return [
            {
                id: 'function',
                title: 'Function',
                dataType: 'gpt-function',
            },
        ];
    }
    getEditors() {
        return [
            {
                type: 'string',
                label: 'Name',
                dataKey: 'name',
            },
            {
                type: 'string',
                label: 'Description',
                dataKey: 'description',
            },
            {
                type: 'code',
                label: 'Schema',
                dataKey: 'schema',
                language: 'json',
            },
        ];
    }
    getBody() {
        return `!markdown_${this.data.name}_: ${this.data.description}`;
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Defines a GPT function, which is a method that the LLM can call in its responses.
      `,
            infoBoxTitle: 'GPT Function Node',
            contextMenuTitle: 'GPT Function',
            group: ['AI'],
        };
    }
    async process(inputs) {
        const parsedSchema = JSON.parse(this.data.schema);
        return {
            ['function']: {
                type: 'gpt-function',
                value: {
                    name: this.data.name,
                    description: this.data.description,
                    parameters: parsedSchema,
                },
            },
        };
    }
}
export const gptFunctionNode = nodeDefinition(GptFunctionNodeImpl, 'GPT Function');
