// @ironclad/rivet-core/dist/esm/model/nodes/MatchNode.js
import { nanoid } from 'nanoid/non-secure';
import { NodeImpl, nodeDefinition } from '../NodeImpl.js';
import { coerceType } from '../../index.js';
import { dedent } from 'ts-dedent';
export class MatchNodeImpl extends NodeImpl {
    static create() {
        const chartNode = {
            type: 'match',
            title: 'Match',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 250,
            },
            data: {
                caseCount: 2,
                cases: ['YES', 'NO'],
            },
        };
        return chartNode;
    }
    getInputDefinitions() {
        const inputs = [
            {
                id: 'input',
                title: 'Test',
                dataType: 'string',
                required: true,
            },
            {
                id: 'value',
                title: 'Value',
                dataType: 'any',
            },
        ];
        return inputs;
    }
    getOutputDefinitions() {
        const outputs = [];
        for (let i = 0; i < this.chartNode.data.caseCount; i++) {
            outputs.push({
                id: `case${i + 1}`,
                title: `Case ${i + 1}`,
                dataType: 'string',
            });
        }
        outputs.push({
            id: 'unmatched',
            title: 'Unmatched',
            dataType: 'string',
        });
        return outputs;
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Any number of regular expressions can be configured, each corresponding to an output of the node. The output port of the first matching regex will be ran, and all other output ports will not be ran.
      `,
            infoBoxTitle: 'Match Node',
            contextMenuTitle: 'Match',
            group: ['Logic'],
        };
    }
    async process(inputs) {
        const inputString = coerceType(inputs['input'], 'string');
        const value = inputs['value'];
        const outputType = value === undefined ? 'string' : value.type;
        const outputValue = value === undefined ? inputString : value.value;
        const cases = this.chartNode.data.cases;
        let matched = false;
        const output = {};
        for (let i = 0; i < cases.length; i++) {
            const regExp = new RegExp(cases[i]);
            const match = regExp.test(inputString);
            if (match) {
                matched = true;
                output[`case${i + 1}`] = {
                    type: outputType,
                    value: outputValue,
                };
            }
            else {
                output[`case${i + 1}`] = {
                    type: 'control-flow-excluded',
                    value: undefined,
                };
            }
        }
        if (!matched) {
            output['unmatched'] = {
                type: outputType,
                value: outputValue,
            };
        }
        else {
            output['unmatched'] = {
                type: 'control-flow-excluded',
                value: undefined,
            };
        }
        return output;
    }
}
export const matchNode = nodeDefinition(MatchNodeImpl, 'Match');
