// @gentrace/core/dist/index.js
'use strict';

var coreApi = require('./api/core-api.js');
var configuration = require('./configuration.js');
var init = require('./providers/init.js');
var evaluation = require('./providers/evaluation.js');
var pipeline = require('./providers/pipeline.js');
var pipelineRun = require('./providers/pipeline-run.js');
var stepRun = require('./providers/step-run.js');
var plugin = require('./providers/plugin.js');
var handler = require('./providers/handler.js');



exports.CoreApi = coreApi.CoreApi;
exports.CoreApiAxiosParamCreator = coreApi.CoreApiAxiosParamCreator;
exports.CoreApiFactory = coreApi.CoreApiFactory;
exports.CoreApiFp = coreApi.CoreApiFp;
exports.Configuration = configuration.Configuration;
Object.defineProperty(exports, 'GENTRACE_API_KEY', {
	enumerable: true,
	get: function () { return init.GENTRACE_API_KEY; }
});
Object.defineProperty(exports, 'GENTRACE_BASE_PATH', {
	enumerable: true,
	get: function () { return init.GENTRACE_BASE_PATH; }
});
Object.defineProperty(exports, 'GENTRACE_BRANCH', {
	enumerable: true,
	get: function () { return init.GENTRACE_BRANCH; }
});
Object.defineProperty(exports, 'GENTRACE_COMMIT', {
	enumerable: true,
	get: function () { return init.GENTRACE_COMMIT; }
});
Object.defineProperty(exports, 'GENTRACE_RUN_NAME', {
	enumerable: true,
	get: function () { return init.GENTRACE_RUN_NAME; }
});
exports.deinit = init.deinit;
Object.defineProperty(exports, 'globalGentraceApi', {
	enumerable: true,
	get: function () { return init.globalGentraceApi; }
});
Object.defineProperty(exports, 'globalGentraceConfig', {
	enumerable: true,
	get: function () { return init.globalGentraceConfig; }
});
exports.init = init.init;
exports.resetGlobalGentraceApi = init.resetGlobalGentraceApi;
exports.constructSubmissionPayload = evaluation.constructSubmissionPayload;
exports.createTestCase = evaluation.createTestCase;
exports.createTestCases = evaluation.createTestCases;
exports.getPipelines = evaluation.getPipelines;
exports.getTestCases = evaluation.getTestCases;
exports.runTest = evaluation.runTest;
exports.submitTestResult = evaluation.submitTestResult;
exports.updateTestCase = evaluation.updateTestCase;
exports.Pipeline = pipeline.Pipeline;
exports.PipelineRun = pipelineRun.PipelineRun;
exports.StepRun = stepRun.StepRun;
exports.GentracePlugin = plugin.GentracePlugin;
exports.isConfig = handler.isConfig;
