// @ironclad/rivet-core/dist/esm/plugins/assemblyAi/LemurTaskNode.js
import { nanoid } from 'nanoid/non-secure';
import { dedent } from 'ts-dedent';
import { coerceTypeOptional, pluginNodeDefinition, } from '../../index.js';
import { getApiKey, getLemurParams, lemurEditorDefinitions, lemurTranscriptIdsInputDefinition } from './lemurHelpers.js';
export const LemurTaskNodeImpl = {
    create() {
        const chartNode = {
            type: 'assemblyAiLemurTask',
            title: 'LeMUR Task',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 250,
            },
            data: {
                final_model: 'default',
            },
        };
        return chartNode;
    },
    getInputDefinitions() {
        return [
            lemurTranscriptIdsInputDefinition,
            {
                id: 'prompt',
                dataType: 'string',
                title: 'Prompt',
            },
        ];
    },
    getOutputDefinitions() {
        return [
            {
                dataType: 'string',
                id: 'response',
                title: 'Response',
            },
        ];
    },
    getEditors() {
        return [
            {
                type: 'string',
                label: 'Prompt',
                dataKey: 'prompt',
            },
            ...lemurEditorDefinitions,
        ];
    },
    getBody() {
        return '';
    },
    getUIData() {
        return {
            infoBoxBody: dedent `Use AssemblyAI LeMUR Custom Task to ask anything.`,
            infoBoxTitle: 'Use AssemblyAI LeMUR Custom Task',
            contextMenuTitle: 'LeMUR Custom Task',
            group: ['AI', 'AssemblyAI'],
        };
    },
    async process(data, inputs, context) {
        const apiKey = getApiKey(context);
        const params = {
            prompt: coerceTypeOptional(inputs['prompt'], 'string') || data.prompt || '',
            ...getLemurParams(inputs, data),
        };
        if (!params.prompt)
            throw new Error('Prompt must be provided.');
        const { response } = await runLemurTask(apiKey, params);
        return {
            ['response']: {
                type: 'string',
                value: response,
            },
        };
    },
};
async function runLemurTask(apiToken, params) {
    const response = await fetch('https://api.assemblyai.com/lemur/v3/generate/task', {
        method: 'POST',
        body: JSON.stringify(params),
        headers: {
            authorization: apiToken,
        },
    });
    const body = await response.json();
    if (response.status !== 200) {
        if ('error' in body)
            throw new Error(body.error);
        throw new Error(`LeMUR Task failed with status ${response.status}`);
    }
    return body;
}
export const lemurTaskNode = pluginNodeDefinition(LemurTaskNodeImpl, 'LeMUR Task');
